### ISSUE TITLE
CASH and LC-ASH report a training-loss artifact in `result.loss`, not the marginal log-likelihood

### ISSUE THREAD

**Issue by yuriisa** [william-denault/cebmf_torch#60]
## Summary

`cash_posterior_means` (and the LC-ASH variants `lcash_posterior_means` / `po_lcash_posterior_means`) return a `loss` field on their result object that is documented and consumed downstream as the negative marginal log-likelihood `-log p(y | prior)`. In the current implementation it is **the sum of per-batch training losses on the final epoch only**, which (a) includes the spike Dirichlet penalty when `penalty > 1`, (b) computes the per-batch term in raw probability space with a `clamp(min=1e-10)` rather than via `logsumexp`, and (c) is therefore not on the same scale as a marginal log-likelihood. Two abandoned `# compute proper full negative marginal log-likelihood` comments in `cebnm/cash_solver.py` confirm this was a planned refactor that never landed.

The same field flows into `cEBMF._update_loadings` / `_update_factors` as `self.kl_l[k] = (-resL.loss) - nm_ll_L`, where the subtraction is only meaningful when `resL.loss == -log p(y | prior)`. Today it is not, so `self.kl_l` is not a KL divergence and `self.obj` is not a lower bound on the ELBO.

## Affected files

- `src/cebmf_torch/cebnm/cash_solver.py`
- `src/cebmf_torch/cebnm/lcash.py`
- `src/cebmf_torch/priors/learned.py` (consumer)
- `src/cebmf_torch/cebmf/cebmf.py` (consumer)

### CASH (`cash_solver.py`)

```python
# lines 203–218: training loop
total_cash_loss = 0.0
for epoch in range(n_epochs):
    epoch_loss = 0.0
    for inputs, targets, noise_std in dataloader:
        ...
        cash_loss = pen_loglik_loss(pred_pi=outputs, marginal_log_lik=batch_loglik, penalty=penalty)
        cash_loss.backward()
        optimizer_cash.step()
        epoch_loss += cash_loss.item()
    total_cash_loss = epoch_loss   # only the LAST epoch is kept
```

```python
# lines 253–263: result construction
 # ---- compute proper full negative marginal log-likelihood (no penalty)

 # ---- compute proper full negative marginal log-likelihood (no penalty)


return cash_PosteriorMeanNorm(
    ...
    loss= total_cash_loss,        # <-- bug
    ...
)
```

The two empty `# ---- compute proper full negative marginal log-likelihood (no penalty)` comments at lines 253 and 255 are dead duplicates from an unfinished refactor.

`pen_loglik_loss` (`cash_solver.py:70–83`) is:

```python
def pen_loglik_loss(pred_pi, marginal_log_lik, penalty=1.5, epsilon=1e-10):
    L_batch = torch.exp(marginal_log_lik)
    inner_sum = torch.sum(pred_pi * L_batch, dim=1)
    inner_sum = torch.clamp(inner_sum, min=epsilon)
    first_sum = torch.sum(torch.log(inner_sum))
    if penalty > 1:
        log_pi0 = torch.log(torch.clamp(pred_pi[:, 0], min=epsilon))
        penalized_log_likelihood_value = first_sum + (penalty - 1) * torch.sum(log_pi0)
    else:
        penalized_log_likelihood_value = first_sum
    return -penalized_log_likelihood_value
```

This is the negative penalized marginal log-likelihood, computed in raw probability space with a `clamp(min=1e-10)`. For an observation whose mixture density underflows to 0, the contribution is hard-clipped to `−log(1e-10) ≈ 23.0` rather than the true (more negative) value. Even when `penalty == 1.0`, the `clamp(min=1e-10)` means the returned scalar still differs from a proper `logsumexp`-based marginal in the small-density tail.

### LC-ASH (`lcash.py`)

```python
# lines 343–359: _train_model
final_epoch_loss = 0.0
for epoch in range(n_epochs):
    epoch_loss = 0.0
    perm = torch.randperm(n, generator=g, device=device)
    for start in range(0, n, batch_size):
        idx = perm[start : start + batch_size]
        pi_pred = model(X_scaled[idx])
        loss = pen_loglik_loss(pi_pred, logL_all[idx], penalty=penalty)
        ...
        epoch_loss += loss.item()
    final_epoch_loss = epoch_loss
return final_epoch_loss
```

```python
# lines 492, 521: _fit_lcash
final_loss = _train_model(...)
...
return cash_PosteriorMeanNorm(
    ...
    loss=final_loss,        # <-- bug
    ...
)
```

LC-ASH inherits the same definition of `pen_loglik_loss` and the same "final-epoch-only" semantics.

## Why it matters

1. **`obj` is not an ELBO.** `cEBMF` (`cebmf/cebmf.py:299, 343`) builds `self.kl_l[k] = (-resL.loss) - nm_ll_L`, then aggregates into `self.obj`. The subtraction only yields a KL when `resL.loss == -log p(y | prior)`. Today it does not, so `self.obj` is not a valid lower bound. Convergence checks based on monotone ELBO are therefore unreliable, and any plot of `self.obj` across iterations is misleading.

2. **Cross-model comparisons are broken.** Two LC-ASH fits with different `penalty` values cannot be compared on `result.loss`, because the penalty term is included. This bites users running model selection (e.g. picking the best penalty), because the scalar reported as the model's loss is on a penalty-dependent scale.

3. **Any downstream workflow that reports `result.loss` is at risk of misinterpretation.** If a summary table, dashboard, or model-selection script reads this field as "−log marginal lik", the reported number is currently on the wrong scale and should not be interpreted that way until this is fixed.

4. **`emdn.py` already does it right.** Lines 288–298 of `cebnm/emdn.py` compute the full marginal log-lik in log space with `logsumexp` after training and store `loss = -full_marginal_ll`. The same template applies trivially to `cash_solver.py` and `lcash.py`.

## Fix

After training, do one full-batch `@torch.no_grad()` forward pass and compute the marginal log-lik in log space:

```python
# pseudo-code, applies to both cash_posterior_means and lcash._fit_lcash
with torch.no_grad():
    pi_full = model(X_scaled_full)                                     # (N, K)
    eps = torch.finfo(pi_full.dtype).tiny
    log_pi = torch.log(pi_full.clamp_min(eps))                         # (N, K)
    data_loglik = get_data_loglik_normal_torch(
        betahat=betahat, sebetahat=sebetahat, location=loc, scale=scale
    )                                                                  # (N, K)
    log_marginal_per_obs = torch.logsumexp(log_pi + data_loglik, dim=1)  # (N,)
    full_marginal_ll = float(log_marginal_per_obs.sum().item())
loss_value = -full_marginal_ll
```

Notes:

- This is the negative marginal log-likelihood **without** the spike Dirichlet penalty. That matches the existing `emdn.py` convention and the meaning required by `cebmf.py`'s KL bookkeeping.
- For LC-ASH this can reuse `_compute_posteriors`-style code: `combined = data_loglik + log_pi_all; log_norm = torch.logsumexp(combined, dim=1)` already lives in `lcash.py:391-393` — the per-observation `log_norm` *is* the per-observation marginal log-lik. We just need to sum it.
- Delete the two dead `# compute proper full negative marginal log-likelihood (no penalty)` comments at `cash_solver.py:253,255`.

## Test gap

There is no test asserting that `result.loss == -log p(y | prior)` on a known-answer toy problem. A minimal regression test would:

1. Generate synthetic data from a known mixture, fit `cash_posterior_means` (and `lcash_posterior_means`) with `penalty=1.0`,
2. Compute the analytic full-data marginal log-likelihood under the *fitted* `pi`, scale, and observation noise via `logsumexp`,
3. Assert `result.loss ≈ -analytic_marginal_loglik` to within numerical tolerance.

This would catch any future regression where the loss field drifts from its documented meaning.
